wizard/src/utils/validation.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { MappedThing, Mapping, MappingType, MappingVal } from "../types";
+
+/*
+ * Checks the provided mapping and returns an array of errors found in it.
+ * On a successful check, the returned array is empty.
+ *
+ * List of checks:
+ * (1) MappingVal should have the right field set depending on MappingType
+ * (2) The Main Three - place, statvar and date - must be mapped
+ * (3) At most one MappedThing can be a COLUMN_HEADER
+ * (4) If there are no COLUMN_HEADER mappings, VALUE should be mapped
+ *     (a) Conversely, if COLUMN_HEADER mapping exists, VALUE must not be mapped
+ * (5) At least one of the mappings must specify a COLUMN or a COLUMN_HEADER
+ * (6) PLACE, when mapped as a COLUMN, must specify placeProperty
+ * (7) VALUE, if it appears, has to be COLUMN
+ *
+ * TODO: Allow placeProperty for all mapping types of PLACE
+ * TODO: Consider returning enum + string pair for error categories
+ *
+ * @param {mappings} CSV mappings
+ * @returns {Array<string>} list of error messages
+ */
+export function checkMappings(mappings: Mapping): Array<string> {
+  const errors = Array<string>();
+  for (const mthing of [
+    MappedThing.PLACE,
+    MappedThing.STAT_VAR,
+    MappedThing.DATE,
+  ]) {
+    if (!mappings.has(mthing)) {
+      // Check #2
+      errors.push("Missing required mapping for " + mthing);
+    }
+  }
+
+  const colHdrThings = Array<string>();
+  let numNonConsts = 0;
+  mappings.forEach((mval: MappingVal, mthing: MappedThing) => {
+    if (mval.type == MappingType.COLUMN) {
```

Comment:
in javascript (/typescript), we usually want to use "===" instead of "=="

Reply:
Ah, thanks, I forgot...